### DISCUSSION TITLE
How does @ModifyVariable work?

### DISCUSSION THREAD

**Discussion by Elmartino4** [FabricMC/fabric-api#1566] [Mod Dev Support]
The [Fabric wiki](https://fabricmc.net/wiki/tutorial:mixin_examples) contains a tutorial on how to use @ModifyVariable. This being;

Mixin:
```
@ModifyVariable(method = "foo()V", at = @At("STORE"), ordinal = 1)
private double injected(double x) {
   return x * 1.5D;
}
```

Result:
```
public void foo() {
   int i0 = doSomething1();
   double d0 = doSomething2();
 - double d1 = doSomething3() + 0.8D;
 + double d1 = injected(doSomething3() + 0.8D);
   double d2 = doSomething4();
}
```

Essentially, I'm unsure how the mixin specified that d1 and only d1 should change rather than d0 or d2.
Any help would be much appretiated!

Also, how do I get variables from within a function that I'm changing another variable in?

**Comment by haykam821** [comment]
The ordinal specifies which of the applicable variables should be modified. Since the mixin method has a `double` parameter and return type, the applicable variables are the ones with the `double` type. In your example, `d1` is the second (or index `1`) of the applicable (`double` type) variables.

You can access other variables in your mixin method by appending them in order to your mixin method's signature.

**Comment by void732** [comment]
More Weird thing...
I had code like this:
```
@Mixin(class1.class)
public class class1Mixin {
	@ModifyVariable(
		method = "foo()V",
		at = @At(value = "STORE", opcode = Opcodes.ASTORE),
		ordinal = 0
	)
	private bar modifyFoo(bar var0) {
		return blah(var0);
	}
```

At the same time in class1 file:
```
public void foo() {
	double a = something1();
	int b = something2();
	- bar anotherFoo = new bar();
	+ bar anotherFoo = modifyFoo(new bar());
	String c = "nothing";
}
```
Yes, seems to be modified smoothly, right?
But, there's always one error in `modifyFoo()` saying: `Expected return type 'class1' for ModifyVariable method.` Wait! Isn't that supposed to be `bar` type???
Fortunately, I CAN compile this code... WTF? And it's functioning well without any bug. Then what was that error all the way?

**Reply by Daomephsta** [reply]
I'm guessing you're using MCDev, and the error shows up when viewing the class in IntelliJ?
Validating Mixins is difficult, so MCDev's validation has many bugs. It's not uncommon for it to generate nonsense errors.
As long as Mixin itself doesn't error, and the decompiler shows the Mixin worked, don't worry about them.

Also please don't provide "example" mixins when asking for help with a specific Mixin. Mixins are usually highly specific to the code they target. If you need help with a specific Mixin, show that Mixin.
